### DISCUSSION TITLE
How does bpf2c handle memory model differences?

### DISCUSSION THREAD

**Discussion by lmb** [microsoft/ebpf-for-windows#3871] [Q&A]
I recently attended a talk by Paul McKenney on the work on a BPF memory model. This exists so that programs accessing memory concurrently have meaningful semantics.

If I understand `bpf2c` correctly it works by translating a BPF instruction into an equivalent C statement. All statements are then compiled via MSVC, resulting in x64 assembly in a native .sys file. In terms of memory models we get the following:

```
Source C (Clang memory model) -> BPF (BPF MM) -> Generated C (MSVC MM) -> assembly (hardware MM)
```

How do you ensure that this transformation of BPF to C preserves BPF memory model semantics in the final assembly?

See https://datatracker.ietf.org/meeting/118/materials/slides-118-bpf-bpf-memory-model-00 for some context.

**Comment by shankarseal** [comment]
Thanks for bringing this up @lmb. Someone from the eBPF maintainers group will look into the presentation and reply back to this discussion.

**Comment by lmb** [comment]
@Alan-Jowett one thought I had: instead of relying on MSVC to lower C to assembly, use the JIT you already have and embed its output as inline assembly in the driver. That removes compiler level reordering from the picture. You still have to worry about the JIT, but that's probably easier to tackle.

**Comment by Alan-Jowett** [comment]
Thanks for the suggestion. This is something we looked into when we originally developed this solution, but this approach would require a lot of additional work.
1) JIT only supports x64 Windows calling conventions, not ARM64 Windows calling conventions.
2) JIT doesn't emit the required debug information.
3) JIT doesn't emit the required pdata and xdata required for correct stack walking.
4) JIT is slower than MSVC generated code.
5) JIT doesn't emit the correct CFG and stack protections that MSVC emits.
6) JIT doesn't emit the correct Spectre mitigations.

In addition, there was a strong pushback from internal folks that we should be using MSVC for native generation so that we would automatically pick-up the latest security mitigations from the compiler.

In short, MSVC gets us machine code with all the latest compiler enhancements and protections that we would need to go and build for the JIT path. This is one of the primary reasons why we are moving towards a native only approach.

**Reply by lmb** [reply]
Thanks for the context! I guess by (1) you mean that there isn't a JIT for arm64?

Would it be possible to take a hybrid approach where each function is jitted individually (kind of what Linux already does) and then the C is something like:

```C
void ebpf_func_1() {
// inline asm
}

void ebpf_func_2() {
// inline asm

ebpf_func_1()

// etc
}
```

So control flow would go via the C compiler and therefore gets CFG integrity, debug metadata, etc. Doesn't solve Spectre, arm64 and speed but would give me a better feeling (for lack of a better word) about the correctness of the emitted code. Maybe I'm seeing ghosts though (ha!)

**Comment by Alan-Jowett** [comment]
Bearing in mind that both compiler and CPU can re-order instructions, I am not sure anything short of explicit memory barriers really make senses long term, especially on architectures like ARM64 which have an even weaker memory model.

There is an ongoing discussion in the BPF mailing list for a proposal for load/acquire and store/release semantics, which I think would be better as it would prevent compiler re-ordering around those instructions.

**Comment by Alan-Jowett** [comment]
@lmb Can we get an explicit example of code that works correctly on Linux but breaks on Windows?

**Reply by lmb** [reply]
Here is an idiom used in eBPF as poor mans atomics:

```
#define __READ_ONCE(x) (*(volatile typeof(x)*)&x)

#define READ_ONCE(x)            \
    ({                          \
        typeof(x) __val;        \
        __val = __READ_ONCE(x); \
        compiler_barrier();     \
        __val;                  \
    })

inline __attribute__((always_inline)) void
compiler_barrier(void)
{
    asm volatile("" ::: "memory");
}

SEC("sample_ext") int test_volatile_read(sample_program_context_t* ctx)
{
    uint32_t key = 0;
    uint32_t* return_value = bpf_map_lookup_elem(&HASH_map, &key);
    if (return_value == NULL) {
        return -1;
    }

    if (READ_ONCE(return_value) != 0) {
        return -2;
    }

    return 0;
}
```

bpf2c generates the following C:
```
#pragma code_seg(push, "sample~1")
static uint64_t
test_volatile_read(void* context, const program_runtime_context_t* runtime_context)
#line 199 "sample/undocked/map.c"
{
#line 199 "sample/undocked/map.c"
    // Prologue.
#line 199 "sample/undocked/map.c"
    uint64_t stack[(UBPF_STACK_SIZE + 7) / 8];
#line 199 "sample/undocked/map.c"
    register uint64_t r0 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r1 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r2 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r3 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r4 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r5 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r6 = 0;
#line 199 "sample/undocked/map.c"
    register uint64_t r10 = 0;

#line 199 "sample/undocked/map.c"
    r1 = (uintptr_t)context;
#line 199 "sample/undocked/map.c"
    r10 = (uintptr_t)((uint8_t*)stack + sizeof(stack));

    // EBPF_OP_MOV64_IMM pc=0 dst=r6 src=r0 offset=0 imm=0
#line 199 "sample/undocked/map.c"
    r6 = IMMEDIATE(0);
    // EBPF_OP_STXW pc=1 dst=r10 src=r6 offset=-4 imm=0
#line 70 "sample/undocked/map.c"
    *(uint32_t*)(uintptr_t)(r10 + OFFSET(-4)) = (uint32_t)r6;
    // EBPF_OP_MOV64_REG pc=2 dst=r2 src=r10 offset=0 imm=0
#line 70 "sample/undocked/map.c"
    r2 = r10;
    // EBPF_OP_ADD64_IMM pc=3 dst=r2 src=r0 offset=0 imm=-4
#line 71 "sample/undocked/map.c"
    r2 += IMMEDIATE(-4);
    // EBPF_OP_LDDW pc=4 dst=r1 src=r1 offset=0 imm=1
#line 71 "sample/undocked/map.c"
    r1 = POINTER(runtime_context->map_data[0].address);
    // EBPF_OP_CALL pc=6 dst=r0 src=r0 offset=0 imm=1
#line 71 "sample/undocked/map.c"
    r0 = runtime_context->helper_data[0].address(r1, r2, r3, r4, r5, context);
#line 71 "sample/undocked/map.c"
    if ((runtime_context->helper_data[0].tail_call) && (r0 == 0)) {
#line 71 "sample/undocked/map.c"
        return 0;
#line 71 "sample/undocked/map.c"
    }
    // EBPF_OP_MOV64_REG pc=7 dst=r1 src=r0 offset=0 imm=0
#line 71 "sample/undocked/map.c"
    r1 = r0;
    // EBPF_OP_LDDW pc=8 dst=r0 src=r0 offset=0 imm=-1
#line 74 "sample/undocked/map.c"
    r0 = (uint64_t)4294967295;
    // EBPF_OP_STXDW pc=10 dst=r10 src=r1 offset=-16 imm=0
#line 74 "sample/undocked/map.c"
    *(uint64_t*)(uintptr_t)(r10 + OFFSET(-16)) = (uint64_t)r1;
    // EBPF_OP_JEQ_IMM pc=11 dst=r1 src=r0 offset=4 imm=0
#line 74 "sample/undocked/map.c"
    if (r1 == IMMEDIATE(0)) {
#line 74 "sample/undocked/map.c"
        goto label_2;
#line 74 "sample/undocked/map.c"
    }
    // EBPF_OP_LDXDW pc=12 dst=r1 src=r10 offset=-16 imm=0
#line 74 "sample/undocked/map.c"
    r1 = *(uint64_t*)(uintptr_t)(r10 + OFFSET(-16));
    // EBPF_OP_JEQ_IMM pc=13 dst=r1 src=r0 offset=1 imm=0
#line 74 "sample/undocked/map.c"
    if (r1 == IMMEDIATE(0)) {
#line 74 "sample/undocked/map.c"
        goto label_1;
#line 74 "sample/undocked/map.c"
    }
    // EBPF_OP_MOV64_IMM pc=14 dst=r6 src=r0 offset=0 imm=-2
#line 74 "sample/undocked/map.c"
    r6 = IMMEDIATE(-2);
label_1:
    // EBPF_OP_MOV64_REG pc=15 dst=r0 src=r6 offset=0 imm=0
#line 74 "sample/undocked/map.c"
    r0 = r6;
label_2:
    // EBPF_OP_EXIT pc=16 dst=r0 src=r0 offset=0 imm=0
#line 75 "sample/undocked/map.c"
    return r0;
#line 199 "sample/undocked/map.c"
}
#pragma code_seg(pop)
#line __LINE__ __FILE__
```

(Side note: I think the line numbers emitted by bpf2c are incorrect? Line 199 in my source is a different function in the same file.)

Neither the volatile read nor the memory barrier are present in the output, because they can't be expressed by the BPF ISA at the moment. You could argue that this goes into undefined behaviour territory: this trick relies on making assumptions about how the JIT maps a 64 bit load into hardware instructions. I think there are two angles to this:

1. The project currently doesn't talk about this at all. From a users perspective it makes sense to assume that "the usual tricks" work, for example the volatile keyword or compiler barriers. Especially when coming from Linux, which arguably has a better story here. The result is going to be very tricky bugs.
2. People do this because having this level of control is useful. My gut feeling is that the BPF ISA won't be able to cover all of them (think control dependencies or similar). Imagine the worst case scenario where people want to start doing RCU in eBPF.
